lib/web_ui/lib/src/engine/text/unicode_range.dart

```diff
@@ -90,3 +108,104 @@ class UnicodePropertyLookup<P> {
+  List<P> propertyEnumValues,
+) {
+  // Packed data is mostly structured in chunks of 9 characters each:
+  //
+  // * [0..3]: Range start, encoded as a base36 integer.
+  // * [4..7]: Range end, encoded as a base36 integer.
+  // * [8]: Index of the property enum value, encoded as a single letter.
+  //
+  // When the range is a single number (i.e. range start == range end), it gets
+  // packed more efficiently in a chunk of 6 characters:
+  //
+  // * [0..3]: Range start (and range end), encoded as a base 36 integer.
+  // * [4]: "!" to indicate that there's no range end.
+  // * [5]: Index of the property enum value, encoded as a single letter.
+
+  // `packedData.length + singleRangesCount * 3` would have been the size of the
+  // packed data if the efficient packing of single-range items wasn't applied.
+  assert((packedData.length + singleRangesCount * 3) % 9 == 0);
+
+  final List<UnicodeRange<P>> ranges = <UnicodeRange<P>>[];
+  int i = 0;
+  while (i < packedData.length) {
+    final int rangeStart = _consumeInt(packedData, i);
+    i += 4;
+
+    int rangeEnd;
+    // "!" <=> 33
+    if (packedData.codeUnitAt(i) == 33) {
+      rangeEnd = rangeStart;
+      i++;
+    } else {
+      rangeEnd = _consumeInt(packedData, i);
+      i += 4;
+    }
+    final int charCode = packedData.codeUnitAt(i);
+    final P property =
+        propertyEnumValues[_getEnumIndexFromPackedValue(charCode)];
+    i++;
+
+    ranges.add(UnicodeRange<P>(rangeStart, rangeEnd, property));
+  }
+  return ranges;
+}
+
+int _getEnumIndexFromPackedValue(int charCode) {
+  // This has to stay in sync with [EnumValue.serialized] in
+  // `tool/unicode_sync_script.dart`.
+
+  // "A" <=> 65
+  // "Z" <=> 90
+  // "a" <=> 97
+  // "z" <=> 122
+
+  assert(
+    (charCode >= 65 && charCode <= 90) || (charCode >= 97 && charCode <= 122),
+  );
+
+  // Uppercase letters were assigned to the first 26 enum values.
+  if (charCode <= 90) {
+    return charCode - 65;
+  }
+  // Lowercase letters were assigned to enum values above 26.
+  return 26 + charCode - 97;
+}
+
+int _consumeInt(String packedData, int index) {
+  // The implementation is equivalent to:
+  //
+  // ```dart
+  // return int.tryParse(packedData.substring(index, index + 4), radix: 36);
+  // ```
+  //
+  // But using substring is slow when called too many times. This custom
+  // implementation makes the unpacking 25%-45% faster than using substring.
+  final int digit0 = _getIntFromCharCode(packedData.codeUnitAt(index + 3));
+  final int digit1 = _getIntFromCharCode(packedData.codeUnitAt(index + 2));
+  final int digit2 = _getIntFromCharCode(packedData.codeUnitAt(index + 1));
+  final int digit3 = _getIntFromCharCode(packedData.codeUnitAt(index));
+  return digit0 + (digit1 * 36) + (digit2 * 36 * 36) + (digit3 * 36 * 36 * 36);
+}
+
+/// Does the same thing as [int.parse(str, 36)] but takes only a single
+/// character as a [charCode] integer.
+int _getIntFromCharCode(int charCode) {
+  // "0" <=> 48
```

Comment:
const int charCode0 = 48, charCodea=... would make code below more readable and you could remove comments.

Reply:
Done